### PR TITLE
Trigger page load listeners when no longer loading

### PR DESCRIPTION
**Pull request by johnhughes1984**
### Description (*)
Trigger JavaScript page load listeners to fire when document is no longer 'loading' rather than waiting until it is marked as 'complete' as this can lead to severe delays in rendering JavaScript content.

### Fixed Issues (if relevant)
1. magento/magento2#22909: requirejs/domReady.js can severely delay rendering of content

### Manual testing scenarios (*)
1. Add a non blocking asset to the page with a severe delay e.g. `` (I added this to `Magento_Theme::html/absolute_footer.phtml`)
2. Visit the homepage
3. The loading status indicator in the browser should be visible for at least the time set in the `sleep` value of the asset, however private / JavaScript content should render before this finishes (I used the `Default welcome msg!` text in the header to verify this)

### Questions or comments
N/A

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=23313) <br/>All committers have signed the CLA.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-5305](https://jira.corp.magento.com/browse/ENGCOM-5305) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by magento-engcom-team** [comment]
@johnhughes1984 thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by ihor-sviziev** [comment]
Not sure if it’s possible to cover with some js tests, but would be really great

**Comment by DrewML** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @DrewML. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @DrewML, here is your [new Magento instance](https://pr-23313.instances.magento-community.engineering).
Admin access: [https://pr-23313.instances.magento-community.engineering/admin](https://pr-23313.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by DrewML** [comment]
I added `` to the homepage content [in the provided demo instance](https://pr-23313.instances.magento-community.engineering/), and verified that:

1. `mage-init` content is no longer relying on the `load` event, but instead `DOMContentLoaded` as intended
2. That the specific conditional change in this PR was actually hit/executed during my testing

Thanks @johnhughes1984!

**Review by DrewML — APPROVED** [review]

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Comment by VladimirZaets** [comment]
Hi @ihor-sviziev. I think test coverage in this case not applicable. In general, we can cover the functionality of domReady  by tests, but it's 3rd party library and it doesn't show real broken cases, we can check it just by functional tests that covered direct part of the functionality

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-5305](https://jira.corp.magento.com/browse/ENGCOM-5305) has been created to process this Pull Request

**Comment by sivaschenko** [comment]
Staging EE functional tests are failing on this PR. The error is related to iframe:

    Element <iframe data-role="preview-iframe" frameborder="0"             
    src="http://extfat.devops.magento.com/? 
    SID=d7161f7bbaad7b4bbfef3ecde303c6b5&amp;___version=1562993400&amp;___store=default" 
    id="preview-iframe" cd_frame_id_="fa616273b878e9e3d2b5ab53b41f17c5">...</iframe> is not 
    clickable at point (132, 111). Other element would receive the click: <div class="actions-toolbar">... 
    </div>

The reason of failed tests is `initialize` of `app/code/Magento/Staging/view/adminhtml/web/js/preview/preview.js` that is triggered before the page content is generated. 

`exports.isIFrameLoaded` is returning `true` even if iFrame content isn't generated. 

The possible solutions are 
 - change the logic of `exports.isIFrameLoaded` to rely on full-page readyState
OR
 - change `exports.initIFrameAjaxInterceptor` to use jQuery object which already included to the file instead of requiring another one https://i.imgur.com/lI2Ig54.png

@VladimirZaets @DrewML can you please advice on the best solution.

**Comment by DrewML** [comment]
@sivaschenko if someone internally can get an EE environment spun up with this branch I'd be happy to help out if someone is unsure what approach to take addressing this.

**Comment by ihor-sviziev** [comment]
Hi @sivaschenko,
Is there any updates? Already 1 months passed since last comment

**Comment by sivaschenko** [comment]
I believe the issue with an iframe is still actual, @VladimirZaets can you please recheck it.

**Comment by davemacaulay** [comment]
We have investigated and provided a fix for this issue as part of https://jira.corp.magento.com/browse/MC-19832. I've sent patches to @johnhughes1984 to be included as part of this PR.

**Comment by ihor-sviziev** [comment]
Hi @VladimirZaets @sivaschenko,
Is there any updates? Is that fix looks good that @davemacaulay provided?

**Comment by johnhughes1984** [comment]
Hi @ihor-sviziev, I've applied the patch provided by @davemacaulay - let me know if you need anything further.

**Comment by davemacaulay** [comment]
@johnhughes1984 @ihor-sviziev this also requires an EE PR which I believe John will be creating.

**Comment by sidolov** [comment]
Hi @johnhughes1984 , will you proceed with PR to EE edition?

**Comment by johnhughes1984** [comment]
Hi @sidolov - yes I'm working on this, this AM - will update once done

**Comment by johnhughes1984** [comment]
@sidolov done - see linked EE PR

**Review by ihor-sviziev — APPROVED** [review]

**Comment by ihor-sviziev** [comment]
So if pr to EE was created - I'm re-approving this PR in order to move it to testing phase

**Comment by ihor-sviziev** [comment]
Hi @engcom-Delta,
Is there any issue with testing of this PR?

**Comment by ihor-sviziev** [comment]
Sorry, closed PR by mistake :( 
I reopened it and assigned all people that were assigned to this PR

**Comment by engcom-Foxtrot** [comment]
@johnhughes1984 please sign CLA.

Without signed new Adobe CLA we not able to process and deliver this PR.

asdasdasdq3123123

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-5305](https://jira.corp.magento.com/browse/ENGCOM-5305) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Review by krzksz — APPROVED** [review]

File: lib/web/requirejs/domReady.js
Comment (andrey-legayev) [inline lib/web/requirejs/domReady.js:92]: There is an issue:
This condition automatically accepts any states which may be introduced in browsers in future.
For example: some browser adds state "initializing" and it breaks Magento2 UI, because it passes this check.

I would change it to following: 
```js
document.readyState === "complete" || document.readyState === "interactive"
```

However it was previously done and caused issues for IE users: https://github.com/requirejs/domReady/issues/1
If Magento2 doesn't support all IE versions then Ok, just add "interactive" as in code above - this is more stable version of logical statement.

BTW Google Chrome also fires "DOMContentLoaded" - it should work fine even without this change.
However, it might be that domReady.js is loaded and executed _after_ DOMContentLoaded fired. It may happen if all static content is cached.
In this case we're waiting for readyState "complete" which makes a problem.
If you saw real improvement then probably this is the real case how it works.

Comment (andrey-legayev) [reply lib/web/requirejs/domReady.js:92]: I've just realized how it works: domReady.js is loaded via RequireJS, this is why it is subscribing to DOMContentLoaded after it fired.
It could be improved if domReady.js added to <head> section statically.

**Comment by ihor-sviziev** [comment]
Just wanted to add links to https://github.com/requirejs/domReady/issues/1 https://github.com/requirejs/domReady/issues/18.
Support of `document.readyState === "interactive"` was removed because of IE9 issue **only**, it's not supported by Magento.

File: lib/web/requirejs/domReady.js
Comment (andrey-legayev) [reply lib/web/requirejs/domReady.js:92]: Me and @ihor-sviziev tested both versions on our project.
Best performance shown when domReady.js is included into <head> section. However you need to modify it's code a bit - add package name to AMD definition:
```js
define("domReady", function(){ ... });
```

Benchmark results on real project with slow 3g mode enabled in Google Chrome dev console:
 - original - first load: 61 sec, second load: 10 sec
 - changes from this pull request - first load: 27 sec, second load: 6.7 sec
 - move domReady.js to `<head>` - first load: 9 sec, second load: 5 sec <== **best result**

Timing was measured by adding following code to `<head>`:
```js
require(["domReady"], function(domReady) {
	domReady(function() {
		console.log(window.performance.now());
	});	
});
```

**Comment by ihor-sviziev** [comment]
Hi,

I'm working on another solution that should fix this issue in the optimal way that @andrey-legayev described above

**Comment by ihor-sviziev** [comment]
Hi,

I created new PR that fixes this issue in more optimal way: https://github.com/magento/magento2/pull/25527

**Comment by DrewML** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @DrewML. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @DrewML, here is your [new Magento instance](https://pr-23313.instances.magento-community.engineering).
Admin access: [https://pr-23313.instances.magento-community.engineering/admin](https://pr-23313.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by DrewML** [comment]
We should move ahead with merging this. The other work being done by @ihor-sviziev in #25527 is additive. This PR is safe, and provides immediate value, so we can merge this while discussing further improvements in the other PR.

File: lib/web/requirejs/domReady.js
Comment (DrewML) [reply lib/web/requirejs/domReady.js:92]: > There is an issue:
This condition automatically accepts any states which may be introduced in browsers in future.
For example: some browser adds state "initializing" and it breaks Magento2 UI, because it passes this check.

See my comment https://github.com/magento/magento2/pull/25527#issuecomment-554101719. Specifically,

> Luckily this isn't something we have to worry much about. This check is an extremely common pattern, and both the WHATWG and W3C have strong requirements to not break the web when adding new features. But it's also a fairly trivial change to update that logic in the other PR, if we did go with it.

The `!== 'loading'` pattern is already how jQuery's `$(document).ready(cb)` function works https://github.com/jquery/jquery/blob/05184cc448f4ed7715ddd6a5d724e167882415f1/src/core/ready.js#L66

**Comment by ihor-sviziev** [comment]
Ok, I don’t mind to merge this in case if jquery has the same check.
I’m approving this pr.
And we’ll discuss my improvements in my PR

**Review by ihor-sviziev — APPROVED** [review]

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Comment by DrewML** [comment]
@magento-engcom-team Anything pending before we can merge this?

**Comment by slavvka** [comment]
@magento run all tests

**Comment by slavvka** [comment]
The PR is currently in the delivery queue. Please be patient.

**Comment by VladimirZaets** [comment]
Please fix failed CE, EE, B2B FAT

**Comment by slavvka** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@engcom-Golf @engcom-Foxtrot , 
Could you fix failing b2b / ee functional tests?

File: lib/web/requirejs/domReady.js
Comment (ihor-sviziev) [reply lib/web/requirejs/domReady.js:92]: @DrewML FYI in the latest version of jQuery it has changed 
https://github.com/jquery/jquery/blob/3.5.1/src/core/ready.js#L67-L72

Maybe we need to change it also

**Pull request merged**
